### DISCUSSION TITLE
App cannot shutdown (ctrl + C) until run_in_threadpool task is finished

### DISCUSSION THREAD

**Discussion by kingbri1** [fastapi/fastapi#11237] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import time
import uvicorn
from fastapi import FastAPI
from fastapi.concurrency import run_in_threadpool

app = FastAPI()

@app.get("/syncsleep")
async def sync_sleep():
    print("Calling sync sleep!")

    # Example of a long-running synchronous process
    # Does not stop until completion even on SIGINT
    await run_in_threadpool(time.sleep, 10000)

if __name__ == "__main__":
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=5001
    )
```

### Description

Hi there! In my project, I'm trying to run a synchronous task in an `async def` route that contains both async and sync code without blocking the main thread. This led me to [this StackOverflow post](https://stackoverflow.com/questions/71516140/fastapi-runs-api-calls-in-serial-instead-of-parallel-fashion) and subsequent issues/discussions which pointed to the awesome `run_in_threadpool` function. 

However, I've encountered an error when force-quitting (via ctrl + c) my application in that it doesn't properly shut down and instead hangs even though uvicorn says the server process is finished. The program is waiting for the background sync task that I spawned via `run_in_threadpool` to finish. Is there a way to make that background thread close on a force shutdown along with the rest of the uvicorn server?

The MRE is an exaggeration of how long the background task will run, but it demonstrates the problem that's occurring.

Reproduction instructions:
1. Start the webserver
2. Call a GET request on `/syncsleep`
3. Wait for a couple of seconds and hit ctrl + c to force quit the application
4. The same traceback in `Additional Context` should fire and the program will hang

Expected result: The application shuts down properly when I force it to quit

Actual result: The application hangs until the long-running background task is finished (or I close the terminal window)

I've added some solutions to the `Additional Context` section. However, these are workarounds and I want to know if there's an easier way to accomplish what I'm looking for. Thanks!

### Operating System

Windows

### Operating System Details

N/A

### FastAPI Version

0.110.0

### Pydantic Version

N/A

### Python Version

3.11.5

### Additional Context

Solution using threads:
```py
import time
import uvicorn
import threading
from fastapi import FastAPI
from fastapi.concurrency import run_in_threadpool

app = FastAPI()

@app.get("/syncsleep")
async def sync_sleep():
    print("Calling sync sleep!")

    # Example of a long-running synchronous process
    # Does not stop until completion even on SIGINT
    await run_in_threadpool(time.sleep, 10000)

def create_webserver():
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=5001
    )

if __name__ == "__main__":
    # Run in a daemon thread to allow for shutdown on SIGINT
    api_thread = threading.Thread(
        target=create_webserver,
        daemon=True
    )

    api_thread.start()
    while api_thread.is_alive():
        time.sleep(0.5)
```

Solution using an abort signal:
```py
import signal
import time
import uvicorn
from fastapi import FastAPI
from fastapi.concurrency import run_in_threadpool
from types import FrameType

app = FastAPI()
abort_signal = False

def sleep_in_loop():
    for _ in range(10000):
        if abort_signal:
            print("Aborting sleep task")
            break

        time.sleep(1)

@app.on_event("startup")
def on_startup():
    default_sigint_handler = signal.getsignal(signal.SIGINT)
    def terminate_now(signum: int, frame: FrameType = None):
        global abort_signal
        abort_signal = True
        default_sigint_handler(signum, frame)
    signal.signal(signal.SIGINT, terminate_now)

@app.get("/syncsleep")
async def sync_sleep():
    print("Calling sync sleep!")

    # Example of a long-running synchronous process
    # Does not stop until completion even on SIGINT
    await run_in_threadpool(sleep_in_loop)

if __name__ == "__main__":
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=5001
    )
```

Traceback from MRE:

```
(venv) D:\python-test> python main.py
INFO:     Started server process [190532]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:5001 (Press CTRL+C to quit)
Calling sync sleep!
10000
INFO:     Shutting down
INFO:     Waiting for connections to close. (CTRL+C to force quit)
INFO:     Finished server process [190532]
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "D:\python-test\venv\Lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 412, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\venv\Lib\site-packages\uvicorn\middleware\proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\venv\Lib\site-packages\fastapi\applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\middleware\errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "D:\python-test\venv\Lib\site-packages\starlette\middleware\exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "D:\python-test\venv\Lib\site-packages\starlette\routing.py", line 758, in __call__
    await self.middleware_stack(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\routing.py", line 778, in app
    await route.handle(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\routing.py", line 299, in handle
    await self.app(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\routing.py", line 79, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "D:\python-test\venv\Lib\site-packages\starlette\_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "D:\python-test\venv\Lib\site-packages\starlette\routing.py", line 74, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\venv\Lib\site-packages\fastapi\routing.py", line 278, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\venv\Lib\site-packages\fastapi\routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\main.py", line 15, in sync_sleep
    await run_in_threadpool(time.sleep, 10000)
  File "D:\python-test\venv\Lib\site-packages\starlette\concurrency.py", line 44, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args, abandon_on_cancel=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\venv\Lib\site-packages\anyio\to_thread.py", line 56, in run_sync
    return await get_async_backend().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python-test\venv\Lib\site-packages\anyio\_backends\_asyncio.py", line 2144, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
asyncio.exceptions.CancelledError
INFO:     127.0.0.1:62346 - "GET /syncsleep HTTP/1.1" 500 Internal Server Error
ERROR:    Traceback (most recent call last):
  File "D:\python-test\venv\Lib\site-packages\starlette\routing.py", line 743, in lifespan
    await receive()
  File "D:\python-test\venv\Lib\site-packages\uvicorn\lifespan\on.py", line 137, in receive
    return await self.receive_queue.get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\bin\miniconda3\Lib\asyncio\queues.py", line 158, in get
    await getter
asyncio.exceptions.CancelledError
```

**Comment by notypecheck** [comment]
Same issue with starlette and uvicorn, using `concurrent.futures.ThreadPoolExecutor` seem to behave the same way too:
```py
import time

import anyio.to_thread
import uvicorn
from starlette.applications import Starlette
from starlette.routing import Route


async def sync_sleep(request):
    print("Calling sync sleep!")
    await anyio.to_thread.run_sync(time.sleep, 10000)


app = Starlette(routes=[Route("/", sync_sleep)])

if __name__ == "__main__":
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=5001
    )
 ```
 ```py
 import time

import anyio.to_thread
import uvicorn


async def app(scope, receive, send):
    assert scope['type'] == 'http'

    print("Calling sync sleep!")
    await anyio.to_thread.run_sync(time.sleep, 10000)
    await send({
        'type': 'http.response.start',
        'status': 200,
        'headers': [
            [b'content-type', b'text/plain'],
        ],
    })
    await send({
        'type': 'http.response.body',
        'body': b'Hello, world!',
    })


if __name__ == "__main__":
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=5001
    )
```

**Comment by Kludex** [comment]
You can't shutdown the thread, this is expected behavior.

**Comment by kingbri1** [comment]
@Kludex That makes sense (see below tests). However, do you mind elaborating on why this behavior is expected? My intuitive understanding is that the entire program will shut down including all background threads because the user wants the program to completely shut down on a force quit.

I tried reproducing this on a def endpoint and the same behavior occurred.
```py
@app.get("/syncsleep")
def sync_sleep():
    print("Sleeping!")
    time.sleep(10000)
```

However, when I kept the endpoint async, I could force-quit with no issues. But then there's the problem of blocking the main thread when the sync task is occurring.
```py
@app.get("/syncsleepreq")
async def sync_sleep_req():
    print("Sleeping!")
    time.sleep(10000)
```

In addition, I decided to try a few more cases and found that trio with hypercorn actually closes all synchronous tasks on SIGINT (example code below with hypercorn from the previous MRE)
```py
import time
import trio
from fastapi import FastAPI
from fastapi.concurrency import run_in_threadpool
from hypercorn import Config
from hypercorn.trio import serve

app = FastAPI()

@app.get("/syncsleep")
async def sync_sleep():
    print("Calling sync sleep!")

    # Example of a long-running synchronous process
    # Stops on SIGINT specifically with trio
    await run_in_threadpool(time.sleep, 10000)

if __name__ == "__main__":
    config = Config()
    config.bind = ["localhost:5001"]
    trio.run(serve, app, config)
 ```

I decided to run this example without FastAPI and just used anyio and the same behavior occurred. Asyncio would hang the application on SIGINT and trio would not.

Asyncio (Hung and shows a KeyboardInterrupt exception):
```py
import time
import asyncio
import anyio
import anyio.to_thread

def blocking_function():
    time.sleep(10000)

async def main():
    print("Sleeping!")
    await anyio.to_thread.run_sync(blocking_function, abandon_on_cancel=True)
    print("Finished")

if __name__ == "__main__":
    asyncio.run(main())
```

Trio:
```py
import time
import trio
import anyio
import anyio.to_thread

def blocking_function():
    time.sleep(10000)

async def main():
    print("Sleeping!")
    await anyio.to_thread.run_sync(blocking_function, abandon_on_cancel=True)
    print("Finished")

if __name__ == "__main__":
    trio.run(main)
```

From there, I did some digging in anyio's code and found that the WorkerThread initialization doesn't set the `daemon` flag to `True` with the asyncio backend [See here](https://github.com/agronholm/anyio/blob/e0529a31e819064fcf1eaf7df55d1712a30bed6b/src/anyio/_backends/_asyncio.py#L802). By setting the daemon flag to true, all threads terminated when the main thread was forcibly quit (via SIGINT). However, I don't know what edge cases will show up from daemonizing `run_in_threadpool` threads.

So, that brings me to a few questions that are a bit confusing.
1. Since `def` request functions have the possibility of hanging, should I make all my requests use `async def`?
2. Why does anyio differ in behavior between asyncio and trio for closing background threads on SIGINT (hanging vs closing the program)?
3. Are the solutions I found in the first comment the correct way to work around this behavior for my use-case?

Thanks in advance.

**Comment by YuriiMotov** [comment]
So, to summarize:
* This is intended behavior of Uvicorn: server tries to terminate the app gracefully
* If you want to force task termination, hit Ctrl+C one more time and third time to finally terminate the server.

**Reply by onionhammer** [reply]
INFO:     Waiting for connections to close. (CTRL+C to force quit)
^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C
